File: src/main/kotlin/leetCode/easy/PalindromeLinkedList.kt
package leetCode.easy

import leetCode.medium.ListNode

class PalindromeLinkedList {
    fun isPalindrome(head: ListNode?): Boolean {
        if (head == null) return true
        if (head.next == null) return true

        val firstHalf = findTheMiddle(head)
        val secondHalf = reverseList(firstHalf.next!!)

        var left: ListNode? = head
        var right: ListNode? = secondHalf
        var result = true
        while (result && right != null) {
            if (left?.`val` != right.`val`) {
                result = false
            }
            left = left?.next
            right = right.next
        }

        firstHalf.next = reverseList(secondHalf)

        return result
    }

    private fun reverseList(head: ListNode): ListNode {
        var prev: ListNode? = null
        var current: ListNode? = head
        while (current != null) {
            val next = current.next
            current.next = prev
            prev = current
            current = next
        }

        return prev!!
    }

    private fun findTheMiddle(head: ListNode): ListNode {
        var fast: ListNode? = head
        var slow: ListNode? = head

        while (fast?.next != null && fast.next?.next != null) {
            fast = fast.next?.next
            slow = slow?.next
        }
        return slow!!
    }
}

fun main(args: Array<String>) {
    println(PalindromeLinkedList().isPalindrome(ListNode(1, ListNode(2, ListNode(2, ListNode(1))))))
    println(PalindromeLinkedList().isPalindrome(ListNode(1, ListNode(2))))
    println(PalindromeLinkedList().isPalindrome(ListNode(1, ListNode(0, ListNode(1)))))
    println(PalindromeLinkedList().isPalindrome(ListNode(1, ListNode(0, ListNode(0)))))
}

File: go/leetCode/easy/PalindromeLinkedList.go
package easy

func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}

	slow := head
	fast := head

	for fast.Next != nil && fast.Next.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
	}
	reversed := innerReverseList(slow.Next)

	left := head
	right := reversed

	for right != nil {
		if left.Val != right.Val {
			return false
		}
		left = left.Next
		right = right.Next
	}

	return true
}

func innerReverseList(head *ListNode) *ListNode {
	var prev *ListNode
	current := head
	for current != nil {
		next := current.Next
		current.Next = prev
		prev = current
		current = next
	}

	return prev
}
